test/hotspot/jtreg/compiler/lib/generators/SpecialDoubleGenerator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.lib.generators;
+
+import java.util.Random;
+import jdk.test.lib.Utils;
+
+/**
+ * Provide a double distribution picked from a list of special values, including NaN, zero, int, etc.
+ */
+public final class SpecialDoubleGenerator extends DoubleGenerator {
+    private static final Random RANDOM = Utils.getRandomInstance();
+
+    // Pre-generated values we can chose from.
+    private static final double[] VALUES = new double[] {
+        0,
+        1,
+        -1,
+        Double.POSITIVE_INFINITY,
+        Double.NEGATIVE_INFINITY,
+        Double.NaN,
+        Double.MAX_VALUE,
+        Double.MIN_NORMAL,
+        Double.MIN_VALUE,
+    };
+
+    // We also mix in other values at a certain percentage.
+    private final DoubleGenerator backgroundGenerator;
+
+    // specialCountDown detemines in how many iterations we generate the next special value.
+    private final int specialMinFrequency;
+    private final int specialMaxFrequency;
+    private int specialCountDown;
+
+    public SpecialDoubleGenerator(DoubleGenerator backgroundGenerator, int specialMinFrequency, int specialMaxFrequency) {
```

Comment:
I think it would be nice to a brief description of what the specialMin/MaxFrequency parameters mean exactly. Also for the float version.

Reply:
Ok, will do.